File: 03_use_cases/13_macos_cua/python/metacua/screenshot.py
"""Screen capture, downscaled to logical points.

A pixel (x, y) in the returned image equals a CGEvent global coordinate (x, y),
so the agent's coordinates map 1:1 onto the cursor position on Retina displays.
"""

import base64
import math
import os
import subprocess
import tempfile
from dataclasses import dataclass
from typing import Optional

from AppKit import (
    NSBitmapImageFileTypePNG,
    NSBitmapImageRep,
    NSColor,
    NSDeviceRGBColorSpace,
    NSGraphicsContext,
    NSImageInterpolationHigh,
    NSRectFill,
    NSScreen,
)
from Foundation import NSData, NSMakeRect, NSMakeSize
from Quartz import CGDisplayBounds, CGMainDisplayID

from .errors import CLIError
from .letterbox import fit_transform


@dataclass
class Screenshot:
    """A captured screen image plus the logical coordinate space it represents."""

    png_base64: str
    width: int  # logical points - the coordinate space the agent operates in
    height: int
    image_width: int
    image_height: int


def primary_screen() -> Optional["NSScreen"]:
    """The primary display whose top-left is the global coordinate origin."""
    screens = NSScreen.screens()
    return screens[0] if screens and len(screens) else None


def _redraw_png(source: "NSBitmapImageRep", width: int, height: int) -> bytes:
    """Redraw a bitmap into a fresh RGBA bitmap of `width`x`height` and PNG-encode it."""
    scaled = NSBitmapImageRep.alloc().initWithBitmapDataPlanes_pixelsWide_pixelsHigh_bitsPerSample_samplesPerPixel_hasAlpha_isPlanar_colorSpaceName_bytesPerRow_bitsPerPixel_(
        None, width, height, 8, 4, True, False, NSDeviceRGBColorSpace, 0, 0,
    )
    if scaled is None:
        raise CLIError("could not allocate a bitmap for scaling")
    scaled.setSize_(NSMakeSize(width, height))

    NSGraphicsContext.saveGraphicsState()
    context = NSGraphicsContext.graphicsContextWithBitmapImageRep_(scaled)
    NSGraphicsContext.setCurrentContext_(context)
    if context is not None:
        context.setImageInterpolation_(NSImageInterpolationHigh)
    source.drawInRect_(NSMakeRect(0, 0, width, height))
    NSGraphicsContext.restoreGraphicsState()

    png = scaled.representationUsingType_properties_(NSBitmapImageFileTypePNG, {})
    if png is None:
        raise CLIError("could not re-encode the screenshot as PNG")
    return bytes(png)


def scaled_letterboxed_png_base64(shot: "Screenshot", width: int, height: int):
    """Aspect-fit `shot` into a black `width`x`height` PNG and return (b64, fit)."""
    fit = fit_transform(shot.width, shot.height, width, height)
    raw_bytes = base64.b64decode(shot.png_base64)
    data = NSData.dataWithBytes_length_(raw_bytes, len(raw_bytes))
    source = NSBitmapImageRep.alloc().initWithData_(data)
    if source is None:
        raise CLIError("could not decode the screenshot for scaling")

    scaled = NSBitmapImageRep.alloc().initWithBitmapDataPlanes_pixelsWide_pixelsHigh_bitsPerSample_samplesPerPixel_hasAlpha_isPlanar_colorSpaceName_bytesPerRow_bitsPerPixel_(
        None, width, height, 8, 4, True, False, NSDeviceRGBColorSpace, 0, 0,
    )
    if scaled is None:
        raise CLIError("could not allocate a bitmap for scaling")
    scaled.setSize_(NSMakeSize(width, height))

    NSGraphicsContext.saveGraphicsState()
    context = NSGraphicsContext.graphicsContextWithBitmapImageRep_(scaled)
    NSGraphicsContext.setCurrentContext_(context)
    if context is not None:
        context.setImageInterpolation_(NSImageInterpolationHigh)
    NSColor.blackColor().setFill()
    # AppKit drawing uses a bottom-left origin in this bitmap context. Centering
    # is symmetric, so the same offsets describe the letterbox transform.
    NSRectFill(NSMakeRect(0, 0, width, height))
    source.drawInRect_(NSMakeRect(fit.ox, fit.oy, fit.drawn_w, fit.drawn_h))
    NSGraphicsContext.restoreGraphicsState()

    png = scaled.representationUsingType_properties_(NSBitmapImageFileTypePNG, {})
    if png is None:
        raise CLIError("could not re-encode the screenshot as PNG")
    return base64.b64encode(bytes(png)).decode("ascii"), fit


def capture_screenshot(scale: float = 1.0) -> Screenshot:
    """Capture the primary display, encoding a possibly downsampled image.

    The full logical display coordinate space is retained for CGEvent actions.
    Uses `screencapture` (requires Screen Recording permission) to avoid the
    deprecated CGDisplayCreateImage path.
    """
    # CGDisplayBounds is thread-safe (this runs on the agent's background thread,
    # where NSScreen would be unsafe). Its size is in logical points, matching the
    # CGEvent global coordinate space the agent operates in.
    bounds = CGDisplayBounds(CGMainDisplayID())
    logical_width = int(round(bounds.size.width))
    logical_height = int(round(bounds.size.height))
    if logical_width <= 0 or logical_height <= 0:
        raise CLIError("no display found to capture")
    if not math.isfinite(scale) or scale <= 0 or scale > 1:
        raise CLIError(f"screenshot scale must be > 0 and <= 1 (got {scale})", code=2)
    image_width = max(1, int(round(logical_width * scale)))
    image_height = max(1, int(round(logical_height * scale)))

    fd, tmp = tempfile.mkstemp(prefix=f"metacua-{os.getpid()}-", suffix="-shot.png")
    os.close(fd)
    try:
        # -x: no capture sound, -t png, -D 1: main display.
        proc = subprocess.run(
            ["/usr/sbin/screencapture", "-x", "-t", "png", "-D", "1", tmp],
            capture_output=True,
        )
        if proc.returncode != 0 or not os.path.exists(tmp):
            raise CLIError(
                f"screencapture failed (status {proc.returncode}). "
                "Grant Screen Recording permission and retry."
            )
        raw = NSData.dataWithContentsOfFile_(tmp)
    finally:
        try:
            os.remove(tmp)
        except OSError:
            pass

    if raw is None:
        raise CLIError("screencapture produced no output; grant Screen Recording permission.")

    source = NSBitmapImageRep.alloc().initWithData_(raw)
    if source is None:
        raise CLIError("could not decode the captured screenshot")

    # Downscale into a fresh RGBA bitmap. The encoded image may be lower
    # resolution than the logical coordinate space used for actions.
    png = _redraw_png(source, image_width, image_height)

    return Screenshot(
        png_base64=base64.b64encode(png).decode("ascii"),
        width=logical_width,
        height=logical_height,
        image_width=image_width,
        image_height=image_height,
    )

File: 03_use_cases/13_macos_cua/src/metacua/Screenshot.swift
import AppKit
import CoreGraphics
import Foundation

/// A captured screen image plus the logical coordinate space it represents.
struct Screenshot {
  let pngBase64: String
  let width: Int // logical points - the coordinate space the agent operates in
  let height: Int
  let imageWidth: Int
  let imageHeight: Int
}

/// The primary display whose top-left is the global coordinate origin.
func primaryScreen() -> NSScreen? {
  NSScreen.screens.first
}

/// Capture the primary display and encode a possibly downsampled image while
/// retaining the full logical display coordinate space for CGEvent actions.
/// Uses `screencapture` (requires Screen Recording permission) to avoid the
/// deprecated CGDisplayCreateImage path.
func captureScreenshot(scale: Double = 1.0) throws -> Screenshot {
  // CGDisplayBounds is thread-safe (this runs on the agent's background thread,
  // where NSScreen would be unsafe). Its size is in logical points, matching the
  // CGEvent global coordinate space the agent operates in.
  let bounds = CGDisplayBounds(CGMainDisplayID())
  let logicalWidth = Int(bounds.width.rounded())
  let logicalHeight = Int(bounds.height.rounded())
  guard logicalWidth > 0, logicalHeight > 0 else {
    throw CLIError("no display found to capture")
  }
  guard scale.isFinite, scale > 0, scale <= 1 else {
    throw CLIError("screenshot scale must be > 0 and <= 1 (got \(scale))", code: 2)
  }
  let imageWidth = max(1, Int((Double(logicalWidth) * scale).rounded()))
  let imageHeight = max(1, Int((Double(logicalHeight) * scale).rounded()))

  let tmp = FileManager.default.temporaryDirectory
    .appendingPathComponent("metacua-\(ProcessInfo.processInfo.processIdentifier)-shot.png")
  defer { try? FileManager.default.removeItem(at: tmp) }

  let process = Process()
  process.executableURL = URL(fileURLWithPath: "/usr/sbin/screencapture")
  // -x: no capture sound, -t png, -D 1: main display.
  process.arguments = ["-x", "-t", "png", "-D", "1", tmp.path]
  let errPipe = Pipe()
  process.standardError = errPipe
  try process.run()
  process.waitUntilExit()
  guard process.terminationStatus == 0, let raw = try? Data(contentsOf: tmp) else {
    throw CLIError(
      "screencapture failed (status \(process.terminationStatus)). "
        + "Grant Screen Recording permission and retry."
    )
  }

  guard let source = NSBitmapImageRep(data: raw) else {
    throw CLIError("could not decode the captured screenshot")
  }

  // Downscale into a fresh RGBA bitmap. The encoded image may be lower
  // resolution than the logical coordinate space used for actions.
  guard
    let scaled = NSBitmapImageRep(
      bitmapDataPlanes: nil,
      pixelsWide: imageWidth,
      pixelsHigh: imageHeight,
      bitsPerSample: 8,
      samplesPerPixel: 4,
      hasAlpha: true,
      isPlanar: false,
      colorSpaceName: .deviceRGB,
      bytesPerRow: 0,
      bitsPerPixel: 0
    )
  else {
    throw CLIError("could not allocate a bitmap for downscaling")
  }
  scaled.size = NSSize(width: imageWidth, height: imageHeight)

  NSGraphicsContext.saveGraphicsState()
  NSGraphicsContext.current = NSGraphicsContext(bitmapImageRep: scaled)
  NSGraphicsContext.current?.imageInterpolation = .high
  source.draw(in: NSRect(x: 0, y: 0, width: imageWidth, height: imageHeight))
  NSGraphicsContext.restoreGraphicsState()

  guard let png = scaled.representation(using: .png, properties: [:]) else {
    throw CLIError("could not re-encode the screenshot as PNG")
  }

  return Screenshot(
    pngBase64: png.base64EncodedString(),
    width: logicalWidth,
    height: logicalHeight,
    imageWidth: imageWidth,
    imageHeight: imageHeight
  )
}

/// Persists screenshots captured during an agent run to `~/.metacua/screenshots/<goalId>/`,
/// so the actual images can be inspected locally instead of only the base64 blobs in traces.
enum ScreenshotStore {
  static var storageURL: URL {
    metacuaHomeURL().appendingPathComponent("screenshots")
  }

  @discardableResult
  static func save(_ screenshot: Screenshot, goalId: String, label: String) throws -> URL {
    guard let data = Data(base64Encoded: screenshot.pngBase64) else {
      throw CLIError("failed to decode screenshot for saving")
    }
    let directory = storageURL.appendingPathComponent(goalId)
    try FileManager.default.createDirectory(
      at: directory,
      withIntermediateDirectories: true,
      attributes: [.posixPermissions: 0o700]
    )
    let url = directory.appendingPathComponent("\(label).png")
    try data.write(to: url, options: [.atomic])
    try? FileManager.default.setAttributes(
      [.posixPermissions: 0o600], ofItemAtPath: url.path)
    return url
  }
}
